src/table_ops.jl

```diff
@@ -0,0 +1,77 @@
+function _write_vals(data, dims::Tuple, perm, missingval)
+    # Allocate Destination Array
+    dst_size = reduce(*, length.(dims))
+    dst = Vector{eltype(data)}(undef, dst_size)
+    dst[perm] .= data
+
+    # Handle Missing Rows
+    _missingval = _cast_missing(data, missingval)
+    missing_rows = ones(Bool, dst_size)
+    missing_rows[perm] .= false
+    return ifelse.(missing_rows, _missingval, dst)
+end
+
+# Find the order of the table's rows according to the coordinate values 
+_sort_coords(table, dims::Tuple) = _sort_coords(_dim_cols(table, dims), dims)
+function _sort_coords(coords::NamedTuple, dims::Tuple)
+    ords = _coords_to_ords(coords, dims)
+    indices = _ords_to_indices(ords, dims)
+    return indices
+end
+
+# Extract coordinate columns from table
+function _dim_cols(table, dims::Tuple)
+    dim_cols = name.(dims)
+    return NamedTuple{dim_cols}(Tables.getcolumn(table, col) for col in dim_cols)
+end
+
+# Extract data columns from table
+function _data_cols(table, dims::Tuple)
+    data_cols = _data_col_names(table, dims)
+    return NamedTuple{Tuple(data_cols)}(Tables.getcolumn(table, col) for col in data_cols)
+end
+
+# Get names of data columns from table
+function _data_col_names(table, dims::Tuple)
+    dim_cols = name.(dims)
+    return filter(x -> !(x in dim_cols), Tables.columnnames(table))
+end
+
+# Determine the ordinality of a set of numerical coordinates
+function _coords_to_ords(coords::AbstractVector, dim::AbstractVector{<:Real})
+    stride = (last(dim) - first(dim)) / (length(dim) - 1)
+    return round.(UInt32, ((coords .- first(dim)) ./ stride) .+ 1)
```

Comment:
Do you mean something like this?

```julia
# Determine the ordinality of a set of regularly spaced numerical coordinates with a starting locus
function _coords_to_ords(
    coords::AbstractVector, 
    dim::Dimension, 
    ::Type{<:Real},
    ::DimensionalData.Start, 
    ::DimensionalData.Regular)
    step = (last(dim) - first(dim)) / (length(dim) - 1)
    return floor.(Int, ((coords .- first(dim)) ./ step) .+ 1)
end

# Determine the ordinality of a set of regularly spaced numerical coordinates with a cenered locus
function _coords_to_ords(
    coords::AbstractVector, 
    dim::Dimension, 
    ::Type{<:Real},
    ::DimensionalData.Center, 
    ::DimensionalData.Regular)
    step = (last(dim) - first(dim)) / (length(dim) - 1)
    return round.(Int, ((coords .- first(dim)) ./ step) .+ 1)
end

# Determine the ordinality of a set of regularly spaced numerical coordinates with an end locus
function _coords_to_ords(
    coords::AbstractVector, 
    dim::Dimension, 
    ::Type{<:Real},
    ::DimensionalData.End, 
    ::DimensionalData.Regular)
    step = (last(dim) - first(dim)) / (length(dim) - 1)
    return ceil.(Int, ((coords .- first(dim)) ./ step) .+ 1)
end

# Determine the ordinality of a set of categorical or irregular coordinates
function _coords_to_ords(coords::AbstractVector, dim::Dimension, ::Any, ::Any, ::Any)
    return map(c -> DimensionalData.selectindices(dim, At(c)), coords)
end

# Preprocessing methods for _coords_to_ords
_coords_to_ords(coords::AbstractVector, dim::Dimension) = _coords_to_ords(coords, dim, eltype(dim), locus(dim), span(dim))
_coords_to_ords(coords::Tuple, dims::Tuple) = Tuple(_coords_to_ords(c, d) for (c, d) in zip(coords, dims))
_coords_to_ords(coords::NamedTuple, dims::Tuple) = _coords_to_ords(Tuple(coords[d] for d in name.(dims)), dims)
```

Reply:
Yep pretty much.

Except use `Contains` for `Intervals` rather than `At`. We could allow users to pass in the selector too, so they can use `Near` or a tolerance with `At`. The default selector can be `nothing` to let use choose specific defaults, or not use selectors at all in the other methods.

(Also note `Tuple(x for x in sometuple)` is not the best pattern for the compiler as it loses the length of the tuple in the iteration, and has to lift it back to a type at runtime. `map` keeps the length so its preferable for small tuples.)

Reply:
I like the idea of letting users choose a selector, but I'm not sure how well it will work in practice. The problem is that `selectindices` is orders of magnitude slower for large regularly spaced dimensions, so it could be a performance footgun. I suppose we could only use it for dimensions caught by the fourth method, but this might result in unexpected behaviour for end users.

Reply:
Sure. But regular At is pretty strict, they need to be the exact values from the lookup.

It should only be one order of magnitude slower for range dims? But yes vector dims will be slow.

Reply:
Here's the results for 3 000 000 coordinates, where Y only has length 1000:

```julia
julia> @btime DimensionalData._coords_to_ords(coords, dims(r,Y), Int, DimensionalData.End(), DimensionalData.Regular());
  3.832 ms (11 allocations: 22.89 MiB)

julia> @btime DimensionalData._coords_to_ords(coords, dims(r,Y), 1, 1, 1);
  949.727 ms (11 allocations: 22.89 MiB)
```

This would take several seconds to reconstruct a 1000 x 1000 array. It would be far slower for anything of a reasonable size.

Reply:
What are the dims?

Reply:
They should be the same speed... Maybe Base.OneTo is missing optimisations for `searchsorted` in base Julia.

In that case we should stop using it!

There shouldn't be a slow path for ranges